[DOCS] Adds loss function description to regression conceptual docs (#1062)
elastic/stack-docs: docs/en/stack/ml/df-analytics/dfa-regression.asciidoc

//end::regression-algorithms[]

[[dfa-regression-lossfunction]]
===== Loss functions for {regression} analyses

> Loss function is for measuring how well a given {ml} model fits the specific 
> data set. The loss function boils down all the different aspects of the model to 
> a single number. The bigger the difference between the prediction and the ground 

Reviewer: ```suggestion
Loss function measures how well a given {ml} model fits the specific 
data set. The loss function boils down all the different under- and overestimations of the model to 
a single number. The bigger the difference between the prediction and the ground 
```

Reviewer: Or something along those lines. Strictly speaking, loss function doesn't care about the properties/aspects of the model, it only takes the predictions and the ground truth into account.

truth, the higher the value of the loss function. Loss functions are used 
automatically in the background during 
<<hyperparameters,hyperparameter optimization>> and when training the decision 
trees to compare the various iterations of the model and decide which iteration 
performs better. 

In the {stack}, there are three different types of loss function:

* https://en.wikipedia.org/wiki/Mean_squared_error[mean squared error (`mse`)]
* mean squared logarithmic error (`msle`; a variation of `mse`)

> * https://en.wikipedia.org/wiki/Huber_loss#Pseudo-Huber_loss_function[Pseudo-Huber loss (`huber`)]

Reviewer: Can we add formulas here? Otherwise, it is unclear what `t` and `delta` mean.

Reviewer: I would also like to add a sentence to the purpose of each loss function. Something like 
MSE: The default choice if no additional information about the data set is available.
MSLE: If you know that the target values are all positive with a long tail distribution (e.g. prices, population, etc.)
Huber: If you want to prevent that the model tries to fit the outliers at the cost of fitting regular data.

> The various types of loss function calculate the model error differently. It 

Reviewer: ```suggestion
The various types of loss function calculate the prediction error differently. It 
```

means that which loss function is appropriate for your case highly depends on 

> the value distribution in your data set, the problem that you want to model, the 

Reviewer: ```suggestion
the target distribution in your data set, the problem that you want to model, the 
```

number of outliers in the data, and so on.

You can determine which loss function to be used during {reganalysis} though the 
{ref}/put-dfanalytics.html[create {dfanalytics-jobs} API]. The default is mean 
squared error (`mse`). If you choose `msle` or `huber`, you can also set up a 
parameter for the loss function which is `t` (defaults to `1.0`) for `msle` and 
`ẟ` (defaults to `1.0`) for `huber`. With these parameters, you can further 
refine the behavior of the chosen functions. There is no parameter to manually 
set up for `mse`.

> TIP: Unless you fully understand the purpose of the different loss functions, it 
> is highly recommended that you leave it unset and use the default parameters.

Reviewer: ```suggestion
TIP: The default loss function parameter values work fine for most of the cases, so unless you fully understand the impact of the different loss function parameters, it is highly recommended that you use the default values.
```

//end::regression-algorithms[]

[[dfa-regression-lossfunction]]
===== Loss functions for {regression} analyses

> Loss function measures how well a given {ml} model fits the specific 

Reviewer 2: I think the noun needs to match the verb here. For example:
```suggestion
A loss function measures how well a given {ml} model fits the specific 
```

> data set. The loss function boils down all the different under- and 

Reviewer 2: Since the previous sentence started with the same phrase, I think you can shorten this one:
```suggestion
data set. It boils down all the different under- and 
```

overestimations of the model to a single number. The bigger the difference 
between the prediction and the ground truth, the higher the value of the loss 
function. Loss functions are used automatically in the background during 
<<hyperparameters,hyperparameter optimization>> and when training the decision 

> trees to compare the various iterations of the model and decide which iteration 

Reviewer 2: This sentence reads quite long, so you might consider shortening it like this:
```suggestion
trees to compare the performance of various iterations of the model.
```

> performs better. 

Reviewer 2: If you accept the suggestion for the first half of this sentence, then this bit is no longer required:
```suggestion

```

In the {stack}, there are three different types of loss function:

* https://en.wikipedia.org/wiki/Mean_squared_error[mean squared error (`mse`)]: 
It is the default choice when no additional information about the data set is 
available.
* mean squared logarithmic error (`msle`; a variation of `mse`): It is for 
cases where the target values are all positive with a long tail distribution 
(for example, prices or population).
* https://en.wikipedia.org/wiki/Huber_loss#Pseudo-Huber_loss_function[Pseudo-Huber loss (`huber`)]:
Use it when you want to prevent the model trying to fit the outliers instead of 

> fit regular data.

Reviewer 2: I don't think it's necessary to repeat "fit" here:
```suggestion
regular data.
```

> The various types of loss function calculate the prediction error differently. 

Reviewer 2: > ... calculate the prediction error differently...

This is the first time we've mentioned the term "prediction error". If this is the appropriate term for the number generated by loss functions, I think it should be introduced earlier. For example in the sentence "The loss function boils down ... to a single number, known as the prediction error".

> It means that which loss function is appropriate for your case highly depends on 

Reviewer 2: IMO this can be shortened:
```suggestion
The appropriate loss function for your use case depends on 
```

the target distribution in your data set, the problem that you want to model, 
the number of outliers in the data, and so on.

> You can determine which loss function to be used during {reganalysis} though the 

Reviewer 2: IMO "determine" usually means "find out" whereas here I think we want to set/specify:
```suggestion
You can specify the loss function to be used during {reganalysis} though the 
```

> {ref}/put-dfanalytics.html[create {dfanalytics-jobs} API]. The default is mean 

Reviewer 2: > {ref}/put-dfanalytics.html[create {dfanalytics-jobs} API].

Will it also be possible to set the loss function in the UI?  If yes, I 
think it's better to just say that you set it when you create the job.

squared error (`mse`). If you choose `msle` or `huber`, you can also set up a 
parameter for the loss function. With the parameter, you can further refine the 

> behavior of the chosen functions. There is no parameter to manually set up for 

Reviewer 2: > There is no parameter to manually...

IMO this sentence is redundant given the way you worded the previous one.
